tools/file_packager.py

```diff
@@ -194,6 +199,121 @@ def add(mode, rootpathsrc, rootpathdst):
     dirnames.extend(new_dirnames)
 
 
+def to_asm_string(string):
+  """Convert a python string to string suitable for including in an
+  assmebly file using the `.asciz` directive.
+
+  The result will be an UTF-8 encoded string in the data section.
+  """
+  # See MCAsmStreamer::PrintQuotedString in llvm/lib/MC/MCAsmStreamer.cpp
+  # And isPrint in llvm/include/llvm/ADT/StringExtras.h
+
+  def is_print(c):
+    return c >= 0x20 and c <= 0x7E
+
+  def escape(c):
+    if is_print(c):
+      return chr(c)
+    escape_chars = {
+      '\b': '\\b',
+      '\f': '\\f',
+      '\n': '\\n',
+      '\r': '\\r',
+      '\t': '\\t',
+    }
+    if c in escape_chars:
+      return escape_chars[c]
+    # Enscode all other chars are three octal digits(!)
+    return '\\%s%s%s' % (oct(c >> 6), oct(c >> 3), oct(c >> 0))
+
+  return ''.join(escape(c) for c in string.encode('utf-8'))
+
+
+def to_c_symbol(filename, used):
+  """Convert a filename (python string) to a legal C symbols, avoiding collisions."""
+  def escape(c):
+     if c.isalnum():
+       return c
+     else:
+       return '_'
+  c_symbol = ''.join(escape(c) for c in filename)
+  # Handle collisions
+  if c_symbol in used:
+    counter = 2
+    while c_symbol + str(counter) in used:
+      counter = counter + 1
+    c_symbol = c_symbol + str(counter)
+  used.add(c_symbol)
+  return c_symbol
+
+
+def generate_object_file(data_files):
+  embed_files = [f for f in data_files if f.mode == 'embed']
+  assert embed_files
+
+  asm_file = shared.replace_suffix(options.obj_output, '.s')
+
+  used = set()
+  ptr_size = 4
```

Comment:
This could be moved down closer to the single use. Also please add a TODO for wasm64.

(Atm we'd get a link error trying to link wasm32 with wasm64, I assume?)